<a href="https://colab.research.google.com/github/ringgoagsya/latihan_python/blob/main/latihan.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# IPhyton Magic Commands

## Pasting Code Blocks: %paste and %cpaste

In [1]:
def donothing(x):
  return x

In [2]:
>>> def donothing(x):
... return x

IndentationError: ignored

kalau langsung copy paste akan salah inden nya. sehingga harus d edit dan di enter setelah mendefinisikan suatu fungsi

In [7]:
%paste
def donothing(x):
  return x
## -- End pasted text --

UsageError: Line magic function `%paste` not found.


magic function %paste tidak ditemukan di library google colab

In [5]:
donothing(10)

10

In [13]:
%cpaste
def donothing(x):
  return x

UsageError: Line magic function `%cpaste` not found.


magic function %paste tidak ditemukan di library google colab

## Running External Code: %run


In [16]:
#-------------------------------------
# file: myscript.py
%%writefile myscript.py
def square(x):
  return x ** 2
for N in range(1, 4):
  print(N, "squared is", square(N))

Overwriting myscript.py


Kodingan diatas untuk membuat file myscript.py dengan kodingan mencari kuadrat dari suatu bilangan dengan syntax X ** 2 artinya x dipangkatkan 2. untuk menjalankan nya menggunakan syntax  %run

In [17]:
%run myscript.py

1 squared is 1
2 squared is 4
3 squared is 9


In [18]:
square(5)

25

untuk menampilkan hasil kuadrat dari suatu bilangan kita bisa memasukkan fungsi square() dan diisi dengan angka yang kita inginkan.

## Timing Code Execution: %timeit


In [20]:
%timeit L = [n ** 2 for n in range(1000)]

1000 loops, best of 5: 313 µs per loop


magic comand  %timeit untuk membandingkan kecepatan waktu eksekusi dari sebuah kode menggunakan list komprehensif.

In [24]:
%%timeit
L = []
for n in range(1000):
  L.append(n ** 2)

1000 loops, best of 5: 382 µs per loop


magic comand  %timeit untuk membandingkan kecepatan waktu eksekusi dari sebuah kode menggunakan for looping.

kita bisa lihat setiap 1000 loop dari list komprehensif membutuhkan waktu 313 microsecond sedangkan untuk loop 382 microsecond setiap loop itu bisa berbeda setiap device. tapi bisa kita lihat bahwa list komprehensif butuh waktu singkat dari looping

## Help on Magic Functions: ?, %magic, and %lsmagic

In [25]:
$timeit?

untuk menampilkan help atau bantuan dari command timeit apa saja yang dapat kita gunakan

In [26]:
%magic

untuk menampilkan help atau bantuan dari command magic apa saja yang dapat kita gunakan

In [27]:
%lsmagic

Available line magics:
%alias  %alias_magic  %autocall  %automagic  %autosave  %bookmark  %cat  %cd  %clear  %colors  %config  %connect_info  %cp  %debug  %dhist  %dirs  %doctest_mode  %ed  %edit  %env  %gui  %hist  %history  %killbgscripts  %ldir  %less  %lf  %lk  %ll  %load  %load_ext  %loadpy  %logoff  %logon  %logstart  %logstate  %logstop  %ls  %lsmagic  %lx  %macro  %magic  %man  %matplotlib  %mkdir  %more  %mv  %notebook  %page  %pastebin  %pdb  %pdef  %pdoc  %pfile  %pinfo  %pinfo2  %pip  %popd  %pprint  %precision  %profile  %prun  %psearch  %psource  %pushd  %pwd  %pycat  %pylab  %qtconsole  %quickref  %recall  %rehashx  %reload_ext  %rep  %rerun  %reset  %reset_selective  %rm  %rmdir  %run  %save  %sc  %set_env  %shell  %store  %sx  %system  %tb  %tensorflow_version  %time  %timeit  %unalias  %unload_ext  %who  %who_ls  %whos  %xdel  %xmode

Available cell magics:
%%!  %%HTML  %%SVG  %%bash  %%bigquery  %%capture  %%debug  %%file  %%html  %%javascript  %%js  %%latex  %%perl 

list magic functuion yang tersedia dan bisa langsung digunakan

# Input and Output History

## 1. IPython's In and Out Objects.

In [28]:
import math

untuk mengimport library dari math. library math berisikan library tentang matematika

In [29]:
math.sin(2)

0.9092974268256817

math.sin() merupakan fungsi untuk menampilkan hasil sinus dari suatu bilangan.

In [30]:
math.cos(2)

-0.4161468365471424

math.cos() merupakan fungsi untuk menampilkan hasil cosinus dari suatu bilangan.

In [31]:
print(In)

['', 'def donothing(x):\n  return x', 'def donothing(x):\nreturn x', "get_ipython().magic('paste')\n>>> def donothing(x):\n... return x\n## -- End pasted text --", "get_ipython().magic('paste')\ndef donothing(x):\n  return x\n## -- End pasted text --", 'donothing(10)', "get_ipython().magic('paste')", "get_ipython().magic('paste')\ndef donothing(x):\n  return x\n## -- End pasted text --", "get_ipython().magic('cpaste')\nPasting code; enter '--' alone on the line to stop or use Ctrl-D.\n:>>> def donothing(x):\n:... return x", "get_ipython().magic('cpaste')\nPasting code; enter '--' alone on the line to stop or use Ctrl-D.\ndef donothing(x):\n  return x", "get_ipython().magic('cpaste --')\ndef donothing(x):\n  return x", '__%cpaste__\ndef donothing(x):\n  return x', 'def __%cpaste__\ndef donothing(x):\n  return x', "get_ipython().magic('cpaste')\ndef donothing(x):\n  return x", 'get_ipython().run_cell_magic(\'writefile\', \'myscript.py\', \'def square(x):\\n  return x ** 2\\nfor N in rang

Menampilkan semua input yang kita input di atas. Yang objeknya digambarkan dengan sebuah list.

In [32]:
Out

{5: 10,
 18: 25,
 27: Available line magics:
%alias  %alias_magic  %autocall  %automagic  %autosave  %bookmark  %cat  %cd  %clear  %colors  %config  %connect_info  %cp  %debug  %dhist  %dirs  %doctest_mode  %ed  %edit  %env  %gui  %hist  %history  %killbgscripts  %ldir  %less  %lf  %lk  %ll  %load  %load_ext  %loadpy  %logoff  %logon  %logstart  %logstate  %logstop  %ls  %lsmagic  %lx  %macro  %magic  %man  %matplotlib  %mkdir  %more  %mv  %notebook  %page  %pastebin  %pdb  %pdef  %pdoc  %pfile  %pinfo  %pinfo2  %pip  %popd  %pprint  %precision  %profile  %prun  %psearch  %psource  %pushd  %pwd  %pycat  %pylab  %qtconsole  %quickref  %recall  %rehashx  %reload_ext  %rep  %rerun  %reset  %reset_selective  %rm  %rmdir  %run  %save  %sc  %set_env  %shell  %store  %sx  %system  %tb  %tensorflow_version  %time  %timeit  %unalias  %unload_ext  %who  %who_ls  %whos  %xdel  %xmode

Available cell magics:
%%!  %%HTML  %%SVG  %%bash  %%bigquery  %%capture  %%debug  %%file  %%html  %%javascript  

menampilkan semua output kita dari awal program. Yang objeknya digambarkan dengan sebuah array

In [34]:
print(In[28])

import math


untuk set input ke berapa kita bisa set In[x]. x disini sesuai angka eksekusinya. misalnya saya ingin menampilkan import math dimana import math saya di eksekusi ke 28 maka saya akan menuliskan In[28] dan untuk mencetaknya kita masukkan ke syntax print -> print(In[28])

In [35]:
print(Out[29])

0.9092974268256817


Fungsi ini dapat digunakan jika kita ingin melihat outpu suatu eksekusi tadi.

In [37]:
Out[29] ** 2 + Out[30] ** 2

1.0

## 2. Underscore Shortcuts and Previous Outputs

In [44]:
print(_)

0.9092974268256817


Shell Python standar hanya memiliki satu pintasan untuk mengakses output sebelumnya yaitu dengan _

In [45]:
print(__)

1.0


untuk mengambil output dari 2 data sebelumnya menggunakan 2 underscore (__)

In [46]:
print(___)

-0.4161468365471424


untuk mengambil output dari 3 data sebelum_nya menggunakan 3 underscore (___)

In [47]:
Out[29]

0.9092974268256817

Untuk custom output kita bisa menggunakan angka eksekusi untuk menampilkan outpu pada eksekusi ke berapa yang kita inginkan. digunakan dengan Out[X] atau _X dimana X adalah nomor baris yang ingin kita akses

In [49]:
_29

0.9092974268256817

## 3. Suppressing Output

In [50]:
math.sin(2) + math.cos(2);

Dari syntax diatas diketahui bahwa hasilnya dihitung secara diam-diam dan hasilnya tidak ditampilkan dilayar atau disimpan dalam Out dictionary. Tetapi datanya sudah disimpan


In [51]:
50 in Out

False

maka dihasilkan hasilnya  "False" berarti ouput dari eksekusi ke 50 belum ada sebelumnnya

## 4. Related Magic Commands

In [53]:
%history -n 28-50

  28: import math
  29: math.sin(2)
  30: math.cos(2)
  31: print(In)
  32: Out
  33: print(In[22])
  34: print(In[28])
  35: print(Out[29])
  36: Out[29] ** 2 + Out[2430] ** 2
  37: Out[29] ** 2 + Out[30] ** 2
  38: print(_)
  39: Out[23]
  40: Out[29]
  41: print(___)
  42: print(__)
  43: _23
  44: print(_)
  45: print(__)
  46: print(___)
  47: Out[29]
  48: _23
  49: _29
  50: math.sin(2) + math.cos(2);


Untuk mengakses sekumpulan input sebelumnya sekaligus, magic functions %history menampilkan hystori kita

# IPython and Shell Commands

## 1. Shell Commands in IPython

In [54]:
!ls

myscript.py  sample_data


In [55]:
!pwd

/content


In [56]:
!echo "printing from the shell"

printing from the shell


Kita dapat menggunakan berbagai command line di IPython dengan menggunakan karakter !

## 2. Passing Values to and from the Shell

In [57]:
contents = !ls

In [58]:
print(contents)

['myscript.py  sample_data']


In [59]:
directory = !pwd

In [60]:
print(directory)

['/content']


seperti diatas cara cara mempassing data menggunakan shell

In [61]:
type(directory)

IPython.utils.text.SList

menampilkan type dari directory dan muncul typenya text.sList

In [62]:
message = "hello from Python"

In [63]:
!echo {message}

hello from Python


menampilkan bisa juga menggunakan !echo() tidak hanya menggunakan print()

# Shell-Related Magic Commands

Jika kita menggunakan IPython shell command dalam beberapa waktu, kita dapat menyadari bahwa tidak bisa menggunakan !cd untuk navigasi file sistem

In [64]:
!pwd

/content


In [65]:
!cd ..

In [66]:
!pwd

/content


Shell command di notebook mengeksekusi dalam sub-shell sementara. Jika kita ingin mengubah direktori kita dapat menggunakan %magic command

In [67]:
%cd

/root


Setelah dicoba kita bisa menggunakannya cd tanpa % 



In [68]:
cd /content

/content


Dimana kita dapat menuliskan command tanpa menggunakan ! dan % seperti untuk fungsi cd,cat,cp,env,ls,man,dll dimana dapat kita gunakan langsung,seperti dilakukan oleh shell biasa. dan biasa dinamakan automagic shell comand

In [69]:
mkdir tmp

In [70]:
ls

myscript.py  sample_data/  tmp/


In [71]:
cp myscript.py tmp/

In [72]:
ls tmp/

myscript.py


In [73]:
rm -r tmp/

Seperti code diatas contoh automagic shell command